<a href="https://colab.research.google.com/github/AmelieSnn/TitanicData/blob/main/EXO_Gouvernance_des_donn%C3%A9es.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# TP 1 Gouvernance des Données


## 1. Exercice : Evaluation et Amélioration de la Qualité des Donnnées


**• Objectif**: Evaluer les problemes de qualite des donnees et mettre
en œuvre des strategies d’amelioration.

**• Source des Donnees**: Utilisez le jeu de donnees ”Titanic” de Titanic - Machine Learning from Disaster, qui comprend des in- formations sur les passagers du Titanic, telles que l’age, la classe, le genre et le statut de survie. vous pouvez utiliser celle de ’train’.

**• Tâches :**

(a) Evaluer le jeu de donnees pour identifier les problemes de qualite, y compris les valeurs manquantes, les doublons, les incoherences et les valeurs aberrantes.

(b) Nettoyer les donnees en gerant les valeurs manquantes, en supprimant les doublons et en corrigeant les incoherences.

(c) Rediger un rapport detaillant les problemes de qualite identifies, les etapes suivies pour nettoyer les donnees, et des recommandations pour la gestion continue de la qualite des donnees.


**• Resultat Attendu :**

– Un jeu de donnees nettoye avec une qualite amelioree.

– Un rapport decrivant les problèmes de qualite des donnees, les actions entreprises et une proposition pour maintenir la qualit ́e des donnees.


## Analyse descriptive préliminaire


In [1]:
# Import des librairies pandas et numpy.
import pandas as pd
import numpy as np

In [3]:
# IMPORTATION DE DRIVE
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
# IMPORTATION DU DATASET (ET SUPPRESSION DE LA COLONNE UNNAMED : 0)
df = pd.read_csv("/content/drive/MyDrive/IASCHOOL/MASTER 2/Gouvernance des données/train.csv", index_col = 'PassengerId')

In [28]:
# Affichage du dataset
df.head(10)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
6                0      0            330877   8.4583   NaN        Q  
7                0      0             17463  51.8625   E46        S  
8                3      1            349909  21.0750   NaN        S  
9                0      2            347742  11.1333   NaN        S  
10               1      0            237736  30.0708   NaN        C

In [14]:
# Informations sur le dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 891 entries, 1 to 891
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Name      891 non-null    object 
 3   Sex       891 non-null    object 
 4   Age       714 non-null    float64
 5   SibSp     891 non-null    int64  
 6   Parch     891 non-null    int64  
 7   Ticket    891 non-null    object 
 8   Fare      891 non-null    float64
 9   Cabin     204 non-null    object 
 10  Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 83.5+ KB


## **(a) Evaluer le jeu de donnees pour identifier les problemes de qualite, y compris les valeurs manquantes, les doublons, les incoherences et les valeurs aberrantes.**

In [15]:
# Les différentes modalités prises par les différentes colonnes

valeurs_uniques = df.apply(lambda col: col.unique())
print(valeurs_uniques)

Survived                                               [0, 1]
Pclass                                              [3, 1, 2]
Name        [Braund, Mr. Owen Harris, Cumings, Mrs. John B...
Sex                                            [male, female]
Age         [22.0, 38.0, 26.0, 35.0, nan, 54.0, 2.0, 27.0,...
SibSp                                   [1, 0, 3, 4, 2, 5, 8]
Parch                                   [0, 1, 2, 5, 3, 4, 6]
Ticket      [A/5 21171, PC 17599, STON/O2. 3101282, 113803...
Fare        [7.25, 71.2833, 7.925, 53.1, 8.05, 8.4583, 51....
Cabin       [nan, C85, C123, E46, G6, C103, D56, A6, C23 C...
Embarked                                       [S, C, Q, nan]
dtype: object


In [8]:
# Statistiques du dataset

df.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [16]:
# Présence de doublons
df.duplicated().sum()

0

In [18]:
# Présence de valeurs manquantes
df.isna().sum()

Survived      0
Pclass        0
Name          0
Sex           0
Age         177
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       687
Embarked      2
dtype: int64

### Detection de valeurs aberrantes (par une analyse contextuelle)

**Survived**

**Analyse** : La colonne Survived est binaire (0 = non survécu, 1 = survécu).

**Min/Max** : min = 0 et max = 1. Ces valeurs sont logiques car il n’y a que deux résultats possibles.

**Conclusion** : Pas de valeurs aberrantes ici, car les données correspondent aux résultats attendus (0 ou 1).

**Pclass**

**Analyse** : La colonne Pclass représente la classe du passager (1ère, 2ème, ou 3ème classe).

**Min/Max** : min = 1 et max = 3. Ces valeurs sont logiques puisqu'il n'y avait que 3 classes à bord.

**Conclusion** : Pas de valeurs aberrantes, car les valeurs sont comprises entre 1 et 3.

**Name**

**Analyse** : Cette colonne contient des chaînes de caractères (noms des passagers). Comme il s'agit d'une donnée textuelle, on ne peut pas détecter de valeurs aberrantes ici avec des statistiques descriptives classiques.

**Conclusion** : Aucune détection automatique de valeurs aberrantes n'est applicable pour les chaînes de caractères.

**Sex**

**Analyse** : La colonne Sex contient des valeurs catégorielles : "male" ou "female".

**Détection d'aberrations**: Ici, on voit qu'il n'existe pas d'autres valeurs que "male" et "female", il n'y a donc pas de valeurs aberrantes

**Conclusion** : Aucune valeur aberrante

**Age**

**Analyse** : La colonne Age représente l'âge des passagers

**Min/Max**: min = 0.42 (âge minimum) et max = 80. La valeur de min = 0.42 pourrait correspondre à un bébé de moins de 6mois, donc elle semble plausible. Le maximum (80 ans) est également plausible.

**Conclusion** : Aucune valeur aberrante



**SibSp**

**Analyse** : La colonne SibSp représente le nombre de personnes de la famille (frères/soeurs ou époux)

**Min/Max**: min = 0 et max = 8. La valeur de min = 0 etle maximum de 8 sont plausibles dans ce contexte.

**Conclusion** : Aucune valeur aberrante



**Parch**

**Analyse** : La colonne Parch représente le nombre de parents ou enfants à bord

**Min/Max**: min = 0 et max = 6. La valeur de min = 0 etle maximum de 8 sont plausibles dans ce contexte.

**Conclusion** : Aucune valeur aberrante

**Ticket**

**Analyse** : Cette colonne contient des numéros de ticket, qui sont des chaînes de caractères. Comme pour la colonne Name, il n'y a pas de méthode statistique pour détecter des valeurs aberrantes.

**Conclusion** : Pas de détection applicable ici.

**Fare**

**Analyse** : Le prix des billets varie énormément, de 0 à 512.3292. La valeur maximale de 512 semble élevée, mais reste plausible pour les cabines de luxe.

**Conclusion** : Nous ne considèrerons aucune valeur aberrante

**Cabin**

**Analyse** : La colonne Cabin contient les numéros de cabine, qui sont des chaînes de caractères. De nombreuses valeurs sont manquantes (NaN). Cela signifie que potentiellement, toutes les personnes n'ont pas de cabine attitrée (peut etre en dortoirs?).

**Conclusion** : Pas de détection applicable ici.

**Embarked**

**Analyse**: La colonne Embarked contient des valeurs catégorielles représentant le port d'embarquement (C = Cherbourg, Q = Queenstown, S = Southampton).

**Détection d'aberrations** : Toutes les valeurs dans cette colonne sont bien "C", "Q" ou "S". Les NaN sont sûrement des oublis.

**Conclusion** : Pas de valeurs aberrante à priori

In [ ]:
################################################################################

## Après ces étapes d'analyse, nous pouvons répondre à la question :


*   **Valeurs manquantes :**

La colonne Age contient des valeurs manquantes (177 valeurs nulles sur 891).
La colonne Cabin a beaucoup de valeurs manquantes (687 valeurs nulles sur 891).
La colonne Embarked a 2 valeurs manquantes.

*   **Doublons :**

Il n'y a pas de doublons dans le dataset

*   **Incohérences :**

Il n'y a pas de valeurs incohérentes

*   **Valeurs aberrantes :**

Après avoir analysé chaque colonne, on ne détecte pas de valeur aberrante

## **(b) Nettoyer les données en gérant les valeurs manquantes, en supprimant les doublons et en corrigeant les incohérences.**




*   **Gérer les valeurs manquantes :**

Pour la colonne Age, on va remplacer les valeurs manquantes par la médiane.

Pour la colonne Embarked, on va remplacer les 2 valeurs manquantes par la valeur la plus fréquente (le mode).

Pour la colonne Cabin, vu le très grand nombre de valeurs manquantes, on va la supprimer, si l'objectif est de faire du machine learning.







In [30]:
# Remplacement des valaurs manquantes par la médiane pour la variable Age.

# Création d'une copie du DataFrame original pour travailler sur df_cleaned
df_cleaned = df.copy()


In [31]:
# Remplacement des valeurs manquantes par la médiane pour la colonne 'Age' dans df_cleaned
df_cleaned['Age'].fillna(df_cleaned['Age'].median(), inplace=True)

In [34]:
# Vérification des modifications
df_cleaned["Age"].isna().sum()

0

In [35]:
# Remplacement des valeurs manquantes par la mode pour la colonne 'Embarked' dans df_cleaned

df_cleaned['Embarked'].fillna(df_cleaned['Embarked'].mode()[0], inplace=True)

In [36]:
# Vérification des modifications
df_cleaned["Embarked"].isna().sum()

0

In [37]:
# Suppression de la colonne 'Cabin' dans df_cleaned
df_cleaned.drop(columns=['Cabin'], inplace=True)

In [38]:
# Vérification des colonnes restantes
print(df_cleaned.columns)

Index(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket',
       'Fare', 'Embarked'],
      dtype='object')



*   **Nettoyage des valeurs aberrantes :**

Il n'y a pas de valeur aberrante à retirer

# **Rapport d’évaluation et d’amélioration de la qualité des données**

## **1. Problèmes de qualité des données identifiés**

L’analyse du jeu de données a révélé plusieurs problèmes de qualité des données qui ont nécessité des actions correctives :

*   **Valeurs manquantes :**

  Trois colonnes principales comportaient des valeurs manquantes :

  La colonne Age comptait 177 valeurs manquantes, soit environ 19.9% du total des passagers. Il est important de noter que l'âge des passagers est une variable essentielle dans de nombreuses analyses (notamment pour prédire les chances de survie).
  La colonne Cabin était particulièrement problématique, avec 687 valeurs manquantes, représentant 77% des données. Cela s'explique probablement par le fait que la majorité des passagers n'ont probalement pas de cabine assignée (notamment ceux en 3ème classe).
  Enfin, la colonne Embarked présentait 2 valeurs manquantes, ce qui, bien que mineur, doit être corrigé pour garantir une analyse complète.

*   **Doublons :**

 Après vérification, il a été constaté qu’il n’y avait aucun doublon dans le dataset, ce qui confirme que chaque ligne représente bien un passager unique.

*   **Incohérences dans les colonnes catégorielles :**
   
  Les valeurs présentes dans les colonnes catégorielles telles que Pclass, Sex, et Embarked ont été vérifiées. Il n'y avait aucune anomalie évidente ou valeur inattendue. Cela indique que ces colonnes sont globalement propres.

*   **Valeurs aberrantes :**
  
  Les colonnes numériques ont également été analysées pour repérer des valeurs aberrantes. Après une analyse contextuelle, aucune des variable ne semble faire ressortir de variable aberrante


## **2. Étapes suivies pour nettoyer les données**

Plusieurs actions correctives ont été mises en œuvre pour améliorer la qualité du dataset :


*   **Traitement des valeurs manquantes :**

  Pour la colonne Age, il a été décidé de remplacer les valeurs manquantes par la médiane des âges, qui s'élève à 28 ans. La médiane a été préférée à la moyenne car elle est moins influencée par les valeurs extrêmes, garantissant ainsi une meilleure représentation des données manquantes.
  Concernant la colonne Embarked, les 2 valeurs manquantes ont été remplacées par la valeur la plus fréquente, à savoir S (Southampton), ce qui permet de conserver la cohérence des données sans introduire de biais.
  Pour la colonne Cabin, en raison du très grand nombre de valeurs manquantes (77% des données), nous avons décidé de supprimer cette colonne. L’information manquante étant trop importante, elle pourrait nuire à la performance d'un modèle de machine learning.

*   **Gestion des valeurs aberrantes :**

  Les âges très élevés et les tarifs inhabituels ont été identifiés et signalés comme des valeurs potentiellement aberrantes. Toutefois, étant donné que ces valeurs peuvent avoir une signification contextuelle (personnes âgées survivantes ou passagers ayant payé des cabines de luxe), elles n’ont pas été supprimées, mais leur impact devra être étudié lors de futures analyses.

*   **Absence de doublons :**
  
  Aucun doublon n'ayant été détecté, aucune action corrective n'a été nécessaire sur ce point.


## **3. Recommandations pour la gestion continue de la qualité des données**

Afin d’assurer une qualité optimale des données à l’avenir, plusieurs recommandations sont proposées :

**Automatiser le nettoyage des données**: Mettre en place des scripts automatisés pour détecter et traiter les valeurs manquantes et les doublons dès la phase d'importation des données. Cela inclut l’utilisation de méthodes d’imputation automatiques pour les colonnes comme Age et Fare.

**Surveillance des valeurs aberrantes :** Il serait utile de mettre en place des mécanismes pour détecter automatiquement les valeurs extrêmes dans les colonnes numériques, afin de signaler les cas où des âges ou des tarifs trop élevés pourraient fausser l'analyse.

**Validation des entrées catégorielles :** Bien que les colonnes catégorielles aient été jugées cohérentes dans ce dataset, un contrôle automatique pour vérifier la cohérence des valeurs dans des colonnes comme Survived, Pclass, Sex, et Embarked devrait être mis en place pour les nouveaux datasets.

En suivant ces recommandations, l'entreprise pourra maintenir une meilleure qualité de données, garantissant ainsi la fiabilité des analyses futures.